tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,362 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Helper functions to add support for iterative magnitude/gradient pruning as seen in the RiGL experiments."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from absl import logging
+
+from tensorflow.python.ops import summary_ops_v2
+from tensorflow.python.summary import summary as summary_ops_v1
+from tensorflow_model_optimization.python.core.sparsity.keras import pruning_utils
+from tensorflow_model_optimization.python.core.sparsity_tf2 import schedule as update_schedule
+from tensorflow_model_optimization.python.core.sparsity_tf2 import pruner
+from tensorflow_model_optimization.python.core.sparsity_tf2 import sparse_utils
+
+
+class RiGLPruner(pruner.Pruner):
+  """
+  Implementation of the RiGL dynamic sparse training algorithm.
+  """
+
+  def __init__(self,
+      update_schedule=update_schedule.ConstantSchedule(0.1, 0),
+      sparsity=0.5,
+      block_size=(1,1),
+      block_pooling_type='AVG',
+      stateless=True,
+      seed=0,
+      noise_std=0,
+      reinit=False,
+      grow_init='zeros'
+    ):
+    """The logic for magnitude-based RiGL trained weight tensors as presented
+    in https://proceedings.icml.cc/static/paper_files/icml/2020/287-Paper.pdf.
+    At each iteration, according to the update_schedule, connections are grown
+    based on gradient information and dropped based on weight magnitude.
+
+    Args:
+      update_schedule: A `PruningSchedule` object that controls pruning rate
+        throughout training. The first arg is fraction of connections to drop
+        during the update, the second is the starting step to begin updating.
+      sparsity: layer specific sparsity at which the dynamic sparse training method uses
+      block_size: The dimensions (height, weight) for the block sparse pattern
+        in rank-2 weight tensors.
+      block_pooling_type: (optional) The function to use to pool weights in the
+        block. Must be 'AVG' or 'MAX'.
+      stateless: whether or not being run on TPU/multi-GPU workers.
+      seed: assigned by PruningConfig (base added) for multiworker consistency in random processes.
+      reinit: boolean for whether to reinitialize a connection that was dropped and regrown to 
+        its original value, or to set it to 0.
+      grow_init: string ('zeros, 'random_normal', 'random_uniform', 'initial_value') way
```

Comment:
Elaborate on this argument and what it actually means. Looking at the code it seems to have something to do w/ how you select the value to assign to newly grown connections?

And does it interact w/ `reinit` in any way? e.g. does it change `0` to something else? Make sure to describe the interactions.